### PR TITLE
Implement Ballerina Constraint Package - Phase 1

### PR DESCRIPTION
**Pull request by ldclakmal**
## Purpose
This PR introduces the Ballerina Constraint package which has APIs to validate the values that have been assigned to Ballerina types.

Proposal: https://github.com/ballerina-platform/ballerina-standard-library/issues/2850
Related issue: https://github.com/ballerina-platform/ballerina-standard-library/issues/2861

This is the first phase of the implementation where this covers followings: https://github.com/ballerina-platform/ballerina-standard-library/issues/2850#issuecomment-1109512766

Also related to https://github.com/ballerina-platform/ballerina-standard-library/issues/3136

## Approach

**Annotation Tags**: `Int`, `Float`, `Number`, `String`, `Array`

**Constraints**:
| annotation | constraints |
|---|---|
| `@constraint:Int` | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive` |
| `@constraint:Float` | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive` |
| `@constraint:Number` | `minValue`, `maxValue`, `minValueExclusive`, `maxValueExclusive` |
| `@constraint:String` | `length`, `minLength`, `maxLength` |
| `@constraint:Array` | `length`, `minLength`, `maxLength` |

```ballerina
public type IntConstraints record {|
   int minValue?;
   int maxValue?;
   int minValueExclusive?;
   int maxValueExclusive?;
|};

public type FloatConstraints record {|
   float minValue?;
   float maxValue?;
   float minValueExclusive?;
   float maxValueExclusive?;
|};

public type NumberConstraints record {|
   decimal minValue?;
   decimal maxValue?;
   decimal minValueExclusive?;
   decimal maxValueExclusive?;
|};

public type StringConstraints record {|
   int length?;
   int minLength?;
   int maxLength?;
|};

public type ArrayConstraints record {|
   int length?;
   int minLength?;
   int maxLength?;
|};
```

**Compatibility**:
| annotation | type |
|---|---|
| `@constraint:Int` | `int` |
| `@constraint:Float` | `float` |
| `@constraint:Number` | `int`\|`float`\|`decimal` |
| `@constraint:String` | `string` |
| `@constraint:Array` | `any[]` |

**Compiler plugin validations**:
1. The annotation with the attached basic data type (since the `type` attachment-point does not validate which type the annotation should attached to).
  ex.: `@constraint:Int` should be attached to `int` type only
  NOTE: The `@Number` annotation can be attached to any subset of `int|float|decimal`.
2. Availability of the constraints of the annotations (since all the record fields are optional).
  ex.: `@constraint:Int {}` shouldn't be declared like this without any constraints
3. Compatibility of the constraints of the annotations (since having some annotations together leads to a contradictory situation)
  ex.: having `length` constraint with `minLength` constraint is a contradiction
4. Validations for the constraints itself (since the constraint package cannot be used to validate the constraint values).
  ex.: The value of the `length` constraint of the `@constraint:String` should be greater than 0.

## Automation tests

- An extraction from the `test_results.json`
    ```json
   "projectName":"constraint",
   "totalTests":333,
   "passed":333,
   "failed":0,
   "skipped":0
    ```

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/~no~
 - Ran FindSecurityBugs plugin and verified report? yes/~no~
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/~no~

## Samples

```ballerina
import ballerina/constraint;

// Use case 1 - Annotations on a record field

type Person record {|
    @constraint:String {
        length: 10
    }
    string name;
|};

public function func1() returns error? {
    Person p = {name: "Alice"};
    p = check constraint:validate(p);
}

// Use case 2 - Annotations on a type

@constraint:Int {
    minValue: 0
}
type Age int;

public function func2() returns error? {
    Age value = 20;
    value = check constraint:validate(value);
}

// Use case 3 - Annotations on a type used as a record field

@constraint:Int {
    maxValue: 100
}
type Weight int;

type Item record {|
    Weight weight;
|};

public function func3() returns error? {
    Item item = {weight: 60};
    item = check constraint:validate(item);
}
```

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/constraint/annotations/TypeAnnotations.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            Type elementType = ((ArrayType) referredType).getElementType();
+            if (elementType instanceof RecordType) {
+                RecordFieldAnnotations recordFieldAnnotations = new RecordFieldAnnotations(this.failedConstraints);
+                for (int i = 0; i < ((BArray) value).getLength(); i++) {
+                    BMap<BString, Object> map = (BMap<BString, Object>) ((BArray) value).getRefValue(i);
+                    recordFieldAnnotations.validate(map, elementType);
+                }
+            } else {
+                // TODO: Validate constraints if element type is a type
+                // for (int i = 0; i < ((BArray) value).getLength(); i++) {
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/constraint/annotations/TypeAnnotations.java]: Are these intentionally commented out?

File: native/src/main/java/io/ballerina/stdlib/constraint/ErrorUtils.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    private static final String VALIDATION_ERROR_MESSAGE_PREFIX = "Validation failed for ";
+    private static final String VALIDATION_ERROR_MESSAGE_SUFFIX = " constraint(s).";
+
+    static BError buildUnexpectedError() {
+        return createError(UNEXPECTED_ERROR_MESSAGE);
+    }
+
+    static BError buildValidationError(Set<String> failedConstraints) {
+        StringBuilder errorMsg = new StringBuilder(VALIDATION_ERROR_MESSAGE_PREFIX);
+        for (String constraint : failedConstraints) {
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/constraint/ErrorUtils.java:42]: Can this `failedConstraints` be empty?

File: native/src/main/java/io/ballerina/stdlib/constraint/Constraints.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        try {
+            Type type = typedesc.getDescribingType();
+            AbstractAnnotations annotations = getAnnotationImpl(type, failedConstraints);
+            annotations.validate(value, type);
+            if (!failedConstraints.isEmpty()) {
+                return ErrorUtils.buildValidationError(failedConstraints);
+            }
+            return null;
+        } catch (RuntimeException e) {
+            return ErrorUtils.buildUnexpectedError();
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/constraint/Constraints.java:50]: Better if we can append the e.getMessage() to the ballerina error message? WDYT?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/AnnotationTagFloat.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public class AnnotationTagFloat implements AnnotationTag {
+
+    @Override
+    public boolean isCompatibleFieldType(String fieldType) {
+        return fieldType.equals(TYPE_FLOAT);
+    }
+
+    @Override
+    public boolean haveCompatibleConstraints(ArrayList<String> constraints) {
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/AnnotationTagFloat.java]: This method seems duplicated among other derived classes. Can we improve the abstraction?

File: native/src/main/java/io/ballerina/stdlib/constraint/annotations/TypeAnnotations.java (hunk 1 above)
Comment (ldclakmal) [reply native/src/main/java/io/ballerina/stdlib/constraint/annotations/TypeAnnotations.java]: Yes. I will fix this or replace the commented lines after having the discussion with runtime team today.

File: native/src/main/java/io/ballerina/stdlib/constraint/ErrorUtils.java (hunk 2 above)
Comment (ldclakmal) [reply native/src/main/java/io/ballerina/stdlib/constraint/ErrorUtils.java:42]: Yes. If all the constraint validations are successful, this `failedConstraints` is empty.

File: native/src/main/java/io/ballerina/stdlib/constraint/annotations/TypeAnnotations.java (hunk 1 above)
Comment (ldclakmal) [reply native/src/main/java/io/ballerina/stdlib/constraint/annotations/TypeAnnotations.java]: Fixed with https://github.com/ballerina-platform/module-ballerina-constraint/pull/1/commits/65e3c937301cc413c316290abad86a07953d39ea

File: native/src/main/java/io/ballerina/stdlib/constraint/Constraints.java (hunk 3 above)
Comment (ldclakmal) [reply native/src/main/java/io/ballerina/stdlib/constraint/Constraints.java:50]: e.getMessage() returns unwanted Java level information to the enduser. This is happened when there is a mismatch between the value and the typedesc. Therefore improved the error message with https://github.com/ballerina-platform/module-ballerina-constraint/pull/1/commits/9c460ac0085962d468987efb96d94e21c337f05e

File: compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/AnnotationTagFloat.java (hunk 4 above)
Comment (ldclakmal) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/AnnotationTagFloat.java]: Fixed with another level of abstraction with https://github.com/ballerina-platform/module-ballerina-constraint/pull/1/commits/27f6a1820945a94165eb0906f558c41e1401b1a1

**Comment by shafreenAnfar** [comment]
Had a call with Chanaka and noticed below things,

- Need to verify against tuple type. For now there should be a compile time failure 
- Error message should be precise but for now it should at least point to the field + failed constraint

The rest looks fine hence we are planning to merge as others are waiting for this. Following PR will address the above points.

**Review by shafreenAnfar — APPROVED** [review]

**Comment by ldclakmal** [comment]
> Had a call with Chanaka and noticed below things,
> 
> * Need to verify against tuple type. For now there should be a compile time failure

This is verified. As of now, there is a compile time failure.
Although we can use referred types as the elements of the tuple as follows, the implementation is blocked due to the runtime API issues.

```ballerina
@constraint:String {
    length: 5
}
type Name string;

@constraint:Int {
    minValue: 18
}
type Age int;

@constraint:Float {
    minValueExclusive: 0
}
type Weight float;

type Person [Name, Age, Weight];
```

> * Error message should be precise but for now it should at least point to the field + failed constraint

The error message is improved as discussed with the PR https://github.com/ballerina-platform/module-ballerina-constraint/pull/10

> 
> The rest looks fine hence we are planning to merge as others are waiting for this. Following PR will address the above points.

**Pull request merged**